flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/strategy/CongestionControlRateLimitingStrategy.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.base.sink.writer.strategy;
+
+import org.apache.flink.util.Preconditions;
+
+/**
+ * A {@code RateLimitingStrategy} implementation that does the following:
+ *
+ * <ul>
+ *   <li>Scales up when any request is successful.
+ *   <li>Scales down when any message in a request is unsuccessful.
+ *   <li>Uses Additive Increase / Multiplicative Decrease (AIMD) strategy to scale up/down.
+ * </ul>
+ *
+ * <p>This strategy works well for throughput-limited record-based sinks (e.g. Kinesis, Kafka).
+ */
+public class CongestionControlRateLimitingStrategy implements RateLimitingStrategy {
+
+    private final int maxInFlightRequests;
+    private final AIMDScalingStrategy aimdScalingStrategy;
```

Comment:
Why can this not be an interface so we can pass in arbitrary scaling strategies? This was part of the original design, which I know was "simplified", but it results in a code smell IMO, We have a coupling here, are we sure this is the right thing to do?

Reply:
I think for this case, we can follow customer requirement, since this coupling is with the CongestionControlStrategy rather than the generic RateLimitingStrategy. At the moment, this CongestionControl strategy is modeled after TCP CongestionControl, which should work really well with AIMD. https://en.wikipedia.org/wiki/TCP_congestion_control